modules/renter/proto/merkleroots.go

```diff
@@ -0,0 +1,343 @@
+	mr.numMerkleRoots--
+	// Truncate file to avoid interpreting trailing data as valid.
+	if err := mr.file.Truncate(rootIndexToOffset(mr.numMerkleRoots)); err != nil {
+		return errors.AddContext(err, "failed to delete last root from file")
+	}
+	// If the last element is uncached we can simply remove it from the slice.
+	if len(mr.uncachedRoots) > 0 {
+		mr.uncachedRoots = mr.uncachedRoots[:len(mr.uncachedRoots)-1]
+		return nil
+	}
+	// If it is not uncached we need to delete the last cached tree and load
+	// its elements into mr.uncachedRoots.
+	mr.cachedSubTrees = mr.cachedSubTrees[:len(mr.cachedSubTrees)-1]
+	rootIndex := len(mr.cachedSubTrees) * (1 << merkleRootsCacheHeight)
+	roots, err := mr.merkleRootsFromIndex(rootIndex, mr.numMerkleRoots)
+	if err != nil {
+		return errors.AddContext(err, "failed to read cached tree's roots")
+	}
+	mr.uncachedRoots = append(mr.uncachedRoots, roots...)
+	return nil
+}
+
+// insert inserts a root by replacing a root at an existing index.
+func (mr *merkleRoots) insert(index int, root crypto.Hash) error {
+	if index > mr.numMerkleRoots {
+		return errors.New("can't insert at a index greater than the number of roots")
+	}
+	if index == mr.numMerkleRoots {
+		return mr.push(root)
+	}
+	// Replaced the root on disk.
+	_, err := mr.file.WriteAt(root[:], rootIndexToOffset(index))
+	if err != nil {
+		return errors.AddContext(err, "failed to insert root on disk")
+	}
+
+	// Find out if the root is in mr.cachedSubTree or mr.uncachedRoots.
+	i, cached := mr.isIndexCached(index)
+	// If the root was not cached we can simply replace it in mr.uncachedRoots.
+	if !cached {
+		mr.uncachedRoots[i] = root
+		return nil
+	}
+	// If the root was cached we need to rebuild the cache.
+	if err := mr.rebuildCachedTree(i); err != nil {
+		return errors.AddContext(err, "failed to rebuild cache for inserted root")
+	}
+	return nil
+}
+
+// isIndexCached determines if the root at index i is already cached in
+// mr.cachedSubTree or if it is still in mr.uncachedRoots. It will return true
+// or false and the index of the root in the corresponding data structure.
+func (mr *merkleRoots) isIndexCached(i int) (int, bool) {
+	if i/(1<<merkleRootsCacheHeight) == len(mr.cachedSubTrees) {
+		// Root is not cached. Return the false and the position in
+		// mr.uncachedRoots
+		return i - len(mr.cachedSubTrees)*(1<<merkleRootsCacheHeight), false
+	}
+	return i / (1 << merkleRootsCacheHeight), true
+}
+
+// lenFromFile returns the number of merkle roots by computing it from the
+// filesize.
+func (mr *merkleRoots) lenFromFile() (int, error) {
+	offset, err := mr.file.Seek(0, io.SeekEnd)
+	if err != nil {
+		return 0, err
+	}
+	// If we haven't written a single root yet we just return 0.
+	rootStart := rootIndexToOffset(0)
+	if offset < rootStart {
+		return 0, nil
+	}
+
+	// Sanity check contract file length.
+	if (offset-rootStart)%crypto.HashSize != 0 {
+		build.Critical("contract file has unexpected length and might be corrupted.")
```

Comment:
this is not deserving of a `build.Critical`. The contract could be truncated to the nearest hash, or it could simply return an error and treat the contract as unusable -- but it's not an unrecoverable error.